# This assignment carries 30 marks. Understand the distribution of the marks - question wise and solve accordingly.

#### Import the required libraries.

In [ ]:
import pandas as pd 
import numpy as np 
import seaborn as sns 
import matplotlib.pyplot as plt 
import warnings 
warnings.filterwarnings('ignore')

#### sklearn library has some inbuilt datasets. We will use the dataset 'Boston'. For that, run the below code to get the dataframe. In this dataframe, 'price' is the target variable.

In [ ]:
from sklearn.datasets import load_boston 
boston = load_boston()
df = pd.DataFrame(boston.data,columns=boston.feature_names)
df['price'] = boston.target
df.head() 

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  price  
0     15.3  396.90   4.98   24.0  
1     17.8  396.90   9.14   21.6  
2     17.8  392.83   4.03   34.7  
3     18.7  394.63   2.94   33.4  
4     18.7  396.90   5.33   36.2

#### 1) Check if any column has null values. (2m)

In [ ]:
df.isnull().sum()

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
price      0
dtype: int64

#### 2) Check the datatypes of all the columns. (2m)

In [ ]:
df.dtypes

CRIM       float64
ZN         float64
INDUS      float64
CHAS       float64
NOX        float64
RM         float64
AGE        float64
DIS        float64
RAD        float64
TAX        float64
PTRATIO    float64
B          float64
LSTAT      float64
price      float64
dtype: object

#### 3)The 'CHAS' column has only two unique entries, 0 and 1 and It is very like to be a categorical column. Change the datatype of this column from float to object.(2m)¶

In [ ]:
df['CHAS'].unique()

array([0., 1.])

In [ ]:
df['CHAS'] = df['CHAS'].astype('object')
df.dtypes

CRIM       float64
ZN         float64
INDUS      float64
CHAS        object
NOX        float64
RM         float64
AGE        float64
DIS        float64
RAD        float64
TAX        float64
PTRATIO    float64
B          float64
LSTAT      float64
price      float64
dtype: object

#### 4) Separate the target variable from independent variables and store it in y. Store independent variables in x. (3m)

In [ ]:
x = df.drop('price',axis = 1)
x

CRIM    ZN  INDUS CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31  0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    0.02731   0.0   7.07  0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729   0.0   7.07  0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237   0.0   2.18  0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905   0.0   2.18  0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
..       ...   ...    ...  ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93  0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527   0.0  11.93  0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076   0.0  11.93  0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959   0.0  11.93  0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741   0.0  11.93  0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       17.8  392.83   4.03  
3       18.7  394.63   2.94  
4       18.7  396.90   5.33  
..       ...     ...    ...  
501     21.0  391.99   9.67  
502     21.0  396.90   9.08  
503     21.0  396.90   5.64  
504     21.0  393.45   6.48  
505     21.0  396.90   7.88  

[506 rows x 13 columns]

In [ ]:
y = df['price']
y

0      24.0
1      21.6
2      34.7
3      33.4
4      36.2
       ... 
501    22.4
502    20.6
503    23.9
504    22.0
505    11.9
Name: price, Length: 506, dtype: float64

#### 5) Split the data into train and test, take 20% of the data in test. (3m)

In [ ]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.2,random_state = 0)

#### 6) Fit the multiple linear regression on the training dataset. (3m)

In [ ]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()
lr.fit(x_train,y_train)

LinearRegression()

#### 6) Get all the coefficients of the regression line. Make a dataframe of column names and respective coefficients. Write down a conclusion for at least one coefficient. (4m)

In [ ]:
lr.coef_ 

array([-1.19443447e-01,  4.47799511e-02,  5.48526168e-03,  2.34080361e+00,
       -1.61236043e+01,  3.70870901e+00, -3.12108178e-03, -1.38639737e+00,
        2.44178327e-01, -1.09896366e-02, -1.04592119e+00,  8.11010693e-03,
       -4.92792725e-01])

In [ ]:
lr.intercept_

38.091694926301855

In [ ]:
coef = pd.DataFrame()
coef['features'] = x_train.columns
coef['coefficients'] = lr.coef_
coef

features  coefficients
0      CRIM     -0.119443
1        ZN      0.044780
2     INDUS      0.005485
3      CHAS      2.340804
4       NOX    -16.123604
5        RM      3.708709
6       AGE     -0.003121
7       DIS     -1.386397
8       RAD      0.244178
9       TAX     -0.010990
10  PTRATIO     -1.045921
11        B      0.008110
12    LSTAT     -0.492793

For a unit change in RAD, the price increases by 0.2441 units, 
for a unit change in RM, the price increases by 3.7087 units, and
for a unit change in ZN, the price increases by 0.044 units.

#### 7) Make predictions on the train data and calculate R2 (R squared). Also write what do you understand by R2 and write the interpretation of this R2 value. (4m)

In [ ]:
train_pred = lr.predict(x_train)
from sklearn.metrics import r2_score
r2 = r2_score(y_train,train_pred)
r2

0.7730135569264233

In [ ]:
lr.score(x_train,y_train)

0.7730135569264233

R^2 is known as the coefficient of determination and is calculated as R^2 = SSR / SST where SST is - Total variation in target column and SSR is - The variation explained by the regression line. The coefficient of determination (R^2) is a number between 0 and 1 that measures how well a statistical model predicts an outcome. Here R^2 = 0.77 which means that out of the total variation in Y, i.e target variable, 77% variation in the target variable can be explained by the independent variables X.

#### 8) Make predictions on the test data and calculate R2 (R squared). (3m)

In [ ]:
test_pred = lr.predict(x_test)
from sklearn.metrics import r2_score
r2_test = r2_score(y_test,test_pred)
r2_test

0.5892223849182526

In [ ]:
lr.score(x_test,y_test)

0.5892223849182526

#### 9) Perform k fold cross validation. Take k = 3. Find out three separate scores. Also find the overall score(average score) (4m)

In [ ]:
from sklearn.model_selection import cross_val_score
scores = cross_val_score(lr,x_train,y_train,cv = 3)
print('Scores are -',scores)
print('Average score is - ',np.mean(scores))

Scores are - [0.75072072 0.68787223 0.73361503]
Average score is -  0.7240693274146016
